client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx

```diff
@@ -0,0 +1,348 @@
+			) {
+				dispatch(
+					recordTracksEvent(
+						'calypso_partner_portal_issue_multiple_licenses_changed_selection_after_dashboard_visit'
+					)
+				);
+			}
+
+			const selectedSiteId = selectedSite?.ID;
+			const fromDashboard = getQueryArg( window.location.href, 'source' ) === 'dashboard';
+
+			// If we need a payment method, redirect now to have the user enter one
+			if ( paymentMethodRequired ) {
+				const nextStep = addQueryArgs(
+					{
+						products: selectedProducts.join( ',' ),
+						...( selectedSiteId && { site_id: selectedSiteId } ),
+						...( fromDashboard && { source: 'dashboard' } ),
+					},
+					partnerPortalBasePath( '/payment-methods/add' )
+				);
+				return page( nextStep );
+			}
+
+			const issueLicenseResponses = await issueLicenses.issueLicenses( selectedProducts );
+
+			const issuedLicenses = issueLicenseResponses.filter(
+				( p ): p is PromiseFulfilledResult< APILicense > => p.status === 'fulfilled'
+			);
+			const issuedKeys = issuedLicenses.map( ( { value } ) => value.license_key );
+			const issuedProducts = issuedKeys
+				.map( ( key ) => {
+					const productSlug = key.split( '_' )[ 0 ];
+					return products?.data?.find( ( p ) => p.slug === productSlug );
+				} )
+				.filter( ( p ): p is APIProductFamilyProduct => p !== undefined );
+
+			// Exit early if we don't see any issued licenses matching a product we know
+			if ( issuedProducts.length === 0 ) {
+				return;
+			}
+
+			dispatch(
+				recordTracksEvent( 'calypso_partner_portal_multiple_licenses_issued', {
+					products: issuedProducts.join( ',' ),
+				} )
+			);
+
+			// If no site is selected, announce that licenses were issued;
+			// then, redirect to somewher more appropriate
+			if ( ! selectedSiteId ) {
+				const issuedProductNames: string[] = issuedProducts.map( ( { name } ) => name );
+				const issuedMessage = getLicenseIssuedMessage( issuedProductNames );
+				dispatch( successNotice( issuedMessage, { displayOnNextPage: true } ) );
+
+				// If this user has no sites, send them to the licenses listing page
+				if ( sitesCount === 0 ) {
+					return page.redirect( partnerPortalBasePath( '/licenses' ) );
+				}
+
+				// If they do have a site, send them to a page where they can assign
+				// the license(s) we just issued
+				const nextStep = addQueryArgs(
+					{
+						products: issuedKeys.join( ',' ),
+					},
+					partnerPortalBasePath( '/assign-license' )
+				);
+
+				return page.redirect( nextStep );
+			}
+
+			// If a specific site is already selected,
+			// let's assign the licenses we just issued to it
+			const assignLicenseResponses = await assignLicensesToSite.assignLicensesToSite( issuedKeys );
+
+			const assignedLicenses = assignLicenseResponses
+				.filter( ( p ): p is PromiseFulfilledResult< APILicense > => p.status === 'fulfilled' )
+				.map( ( { status, value } ) => {
+					const productSlug = value.license_key.split( '_' )[ 0 ];
```

Comment:
A while ago I realized that siteless licenses actually can include multiple underscores in the license key so here we should be splitting on the last underscore instead of the first.
Also, if we don't have something in place already, we should move this to a reusable function so we don't have magic license key string manipulation logic in multiple places.

Reply:
cc @yashwin, as I believe you worked on this, and we discussed it at some point.

Reply:
Yes. We had a similar discussion where splitting the license key could be an issue, but I think I missed doing that here.

I have created a [ticket](1202619025189113-as-1205001175945323) with more info.